orbit/pkg/table/tcc_access/tcc_access.go

```diff
@@ -0,0 +1,179 @@
+	sqlite3Path   = "/usr/bin/sqlite3"
+	dbColNames    = []string{"service", "client", "client_type", "auth_value", "auth_reason", "last_modified", "policy_id", "indirect_object_identifier", "indirect_object_identifier_type"}
+)
+
+// Columns is the schema of the table.
+func Columns() []table.ColumnDefinition {
+	return []table.ColumnDefinition{
+		// added here
+		table.TextColumn("source"),
+		table.IntegerColumn("uid"),
+		// derived from a TCC.db
+		table.TextColumn("service"),
+		table.TextColumn("client"),
+		table.IntegerColumn("client_type"),
+		table.IntegerColumn("auth_value"),
+		table.IntegerColumn("auth_reason"),
+		table.BigIntColumn("last_modified"),
+		table.IntegerColumn("policy_id"),
+		table.TextColumn("indirect_object_identifier"),
+		table.IntegerColumn("indirect_object_identifier_type"),
+	}
+}
+
+// Generate is called to return the results for the table at query time.
+// Constraints for generating can be retrieved from the queryContext.
+
+func Generate(ctx context.Context, queryContext table.QueryContext) ([]map[string]string, error) {
+	// get all usernames on the mac
+	cmd := exec.Command("dscl", ".", "list", "/Users")
+	out, err := cmd.Output()
+	if err != nil {
+		return nil, err
+	}
+	allUsernames := strings.Split(string(out[:]), "\n")
+	var usernames []string
+	for _, username := range allUsernames {
+		if !strings.HasPrefix(username, "_") && username != "nobody" && username != "root" && username != "daemon" {
+			usernames = append(usernames, username)
+		}
+	}
+
+	if testContext {
+		usernames = []string{"testUser1", "testUser2"}
+	}
+
+	// build rows for every user-level TCC.db
+	var rows []map[string]string
+	for _, username := range usernames {
+		uRs, err := getTCCAccessRows(username)
+		if err != nil {
+			return nil, err
+		}
+		rows = append(rows, uRs...)
+	}
+
+	// and for the system-level TCC.db
+	sRs, err := getTCCAccessRows("")
+	if err != nil {
+		return nil, err
+	}
+	rows = append(rows, sRs...)
+
+	return rows, nil
+}
+
+func getTCCAccessRows(username string) ([]map[string]string, error) {
+	// avoids additional C compilation requirements that would be introduced by using
+	// https://github.com/mattn/go-sqlite3
+
+	tccPath := tccPathPrefix
+	if username != "" {
+		tccPath += "/Users/" + username
+	}
+	tccPath += tccPathSuffix
+	cmd := exec.Command(sqlite3Path, tccPath, dbQuery)
+	var dbOut bytes.Buffer
+	var stderr bytes.Buffer
+	cmd.Stdout = &dbOut
+	cmd.Stderr = &stderr
+	err := cmd.Run()
+	if err != nil {
+		return nil, fmt.Errorf("Generate failed at `cmd.Run()`:"+stderr.String()+":%w", err)
+	}
+
+	parsedRows := parseTCCDbReadOutput(dbOut.Bytes())
+
+	rows, err := buildTableRows(username, parsedRows)
+	if err != nil {
+		return nil, err
+	}
+
+	return rows, nil
+}
+
+func parseTCCDbReadOutput(dbOut []byte) [][]string {
+	// split by newLine for rows, then by "|" for columns
+	rawRows := strings.Split(string(dbOut[:]), "\n")
+	n := len(rawRows)
+	for len(rawRows[n-1]) == 0 {
```

Comment:
Why a for loop, can there be multiple rows with no content (just `\n`)?

Reply:
It doesn't seem like there can be more than one `\n`, and I never observed more than one, but figured this would cover it in case it did occur. Better to just `[:n-1]` it you think?

Reply:
Yeah, no loop is better IMO.